File: python/webapp.py
import os
import datetime
from flask import Flask, render_template, request, redirect, url_for, flash
from core.database import ByteForceDB

app = Flask(__name__)
app.secret_key = 'byteforce-4fefde4566-tech-assssment-54ftsdrtse456-secret-key'

# Initialize Database
db = ByteForceDB(data_dir="data")

def init_db():
    tables = db.storage.list_tables()
    
    if 'tasks' not in tables:
        db.execute("""
            CREATE TABLE tasks (
                id INTEGER PRIMARY KEY, 
                title TEXT, 
                status TEXT, 
                created_at TEXT, 
                is_draft BOOLEAN,
                is_deleted BOOLEAN
            )
        """)
    
    if 'subtasks' not in tables:
        # distinct column names to avoid collision in simple dict-based join
        db.execute("""
            CREATE TABLE subtasks (
                sid INTEGER PRIMARY KEY,
                parent_id INTEGER,
                subtitle TEXT,
                substatus TEXT
            )
        """)
        print("Database initialized with tasks and subtasks.")

init_db()

def get_next_id(table_name, id_col):
    rows = db.execute(f"SELECT * FROM {table_name}")
    if isinstance(rows, list) and rows:
        return max([r.get(id_col, 0) for r in rows]) + 1
    return 1

@app.route('/')
def index():
    return redirect(url_for('list_tasks', view='active'))

@app.route('/tasks/<view>')
def list_tasks(view):
    # USE THE NEW LEFT JOIN!
    # Fetch tasks and their subtasks in one go
    raw_data = db.execute("SELECT * FROM tasks LEFT JOIN subtasks ON id = parent_id")
    
    if isinstance(raw_data, str) and raw_data.startswith("Error"):
        return raw_data, 500
    
    # Process flat join results into hierarchical objects
    tasks_map = {}
    
    for row in raw_data:
        t_id = row.get('id')
        if not t_id: continue # Should not happen unless row is empty
        
        if t_id not in tasks_map:
            # New task encountered
            tasks_map[t_id] = {
                'id': row['id'],
                'title': row['title'],
                'status': row['status'],
                'created_at': row['created_at'],
                'is_draft': row.get('is_draft', False),
                'is_deleted': row.get('is_deleted', False),
                'subtasks': []
            }
        
        # Check if this row has a subtask (sid will be present and not None)
        if row.get('sid') is not None:
            subtask = {
                'sid': row['sid'],
                'parent_id': row['parent_id'],
                'subtitle': row['subtitle'],
                'substatus': row['substatus']
            }
            tasks_map[t_id]['subtasks'].append(subtask)

    # Filter based on view
    final_tasks = []
    for task in tasks_map.values():
        is_deleted = task['is_deleted']
        is_draft = task['is_draft']
        
        if view == 'trash':
            if is_deleted: final_tasks.append(task)
        elif view == 'drafts':
            if not is_deleted and is_draft: final_tasks.append(task)
        else: # active
            if not is_deleted and not is_draft: final_tasks.append(task)

    # Sort
    final_tasks.sort(key=lambda x: (x.get('status') == 'completed', -x.get('id')))
    
    return render_template('index.html', tasks=final_tasks, view=view)

@app.route('/add', methods=['POST'])
def add():
    title = request.form.get('title')
    is_draft = request.form.get('is_draft') == 'on'
    
    if title:
        new_id = get_next_id('tasks', 'id')
        created_at = datetime.datetime.now().strftime("%Y-%m-%d %H:%M")
        db.execute(
            "INSERT INTO tasks VALUES (?, ?, 'pending', ?, ?, ?)", 
            [new_id, title, created_at, is_draft, False]
        )
        flash("Draft saved." if is_draft else "Task added.", 'success')
        
    return redirect(url_for('list_tasks', view='drafts' if is_draft else 'active'))

@app.route('/subtask/add', methods=['POST'])
def add_subtask():
    parent_id = request.form.get('parent_id')
    subtitle = request.form.get('subtitle')
    
    if parent_id and subtitle:
        sid = get_next_id('subtasks', 'sid')
        db.execute(
            "INSERT INTO subtasks VALUES (?, ?, ?, 'pending')",
            [sid, int(parent_id), subtitle]
        )
        flash("Subtask added.", "success")
        
    return redirect(url_for('list_tasks', view='active'))

@app.route('/subtask/toggle/<int:sid>')
def toggle_subtask(sid):
    # We need to find the subtask first to check status
    # Simple query (no join needed for simple update)
    rows = db.execute("SELECT * FROM subtasks WHERE sid = ?", [sid])
    if rows and isinstance(rows, list):
        st = rows[0]
        new_status = 'completed' if st['substatus'] == 'pending' else 'pending'
        db.execute("UPDATE subtasks SET substatus = ? WHERE sid = ?", [new_status, sid])
    
    return redirect(url_for('list_tasks', view='active'))

@app.route('/action/<action>/<int:task_id>')
def task_action(action, task_id):
    rows = db.execute("SELECT * FROM tasks WHERE id = ?", [task_id])
    if not rows or not isinstance(rows, list): return "Task not found", 404
    task = rows[0]
    
    if action == 'toggle':
        new_status = 'completed' if task['status'] == 'pending' else 'pending'
        db.execute("UPDATE tasks SET status = ? WHERE id = ?", [new_status, task_id])
    elif action == 'delete':
        db.execute("UPDATE tasks SET is_deleted = ? WHERE id = ?", [True, task_id])
        flash('Moved to Trash.', 'warning')
    elif action == 'restore':
        db.execute("UPDATE tasks SET is_deleted = ? WHERE id = ?", [False, task_id])
        flash('Task restored.', 'success')
    elif action == 'publish':
        db.execute("UPDATE tasks SET is_draft = ? WHERE id = ?", [False, task_id])
        flash('Task published!', 'success')
    elif action == 'destroy':
        db.execute("DELETE FROM tasks WHERE id = ?", [task_id])
        # Also clean up subtasks? ideally yes, but keeping it simple
        db.execute("DELETE FROM subtasks WHERE parent_id = ?", [task_id])
        flash('Permanently deleted.', 'danger')

    referer = request.headers.get("Referer", "")
    if 'trash' in referer: return redirect(url_for('list_tasks', view='trash'))
    if 'drafts' in referer: return redirect(url_for('list_tasks', view='drafts'))
    return redirect(url_for('list_tasks', view='active'))

if __name__ == '__main__':
    app.run(debug=True, port=5000)

File: java/src/main/java/com/byteforce/web/WebApp.java
package com.byteforce.web;

import static spark.Spark.*;

import com.byteforce.core.ByteForceDB;
import com.byteforce.core.ExecutionResult;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.stream.Collectors;
import spark.ModelAndView;
import spark.template.handlebars.HandlebarsTemplateEngine;

public class WebApp {
  private static ByteForceDB db;

  public static void main(String[] args) {
    port(4567);

    db = new ByteForceDB("data");
    initDb();

    HandlebarsTemplateEngine engine = new HandlebarsTemplateEngine();

    get(
        "/",
        (req, res) -> {
          res.redirect("/tasks/active");
          return null;
        });

    get(
        "/tasks/:view",
        (req, res) -> {
          String view = req.params(":view");
          ExecutionResult result =
              db.execute("SELECT * FROM tasks LEFT JOIN subtasks ON id = parent_id");

          if (result.isError()) return "Error: " + result.getMessage();

          // Intermediate structure to group subtasks by task ID
          Map<Integer, TaskBuilder> builders = new LinkedHashMap<>();
          for (Map<String, Object> row : result.getRows()) {
            Integer id = (Integer) row.get("id");
            if (id == null) continue;

            builders.putIfAbsent(id, new TaskBuilder(row));
            TaskBuilder builder = builders.get(id);

            if (row.get("sid") != null) {
              builder.addSubTask(row);
            }
          }

          List<Task> finalTasks =
              builders.values().stream()
                  .map(TaskBuilder::build)
                  .filter(
                      t -> {
                        if ("trash".equals(view)) return t.isDeleted();
                        if ("drafts".equals(view)) return !t.isDeleted() && t.isDraft();
                        return !t.isDeleted() && !t.isDraft();
                      })
                  .sorted(
                      (t1, t2) -> {
                        if (t1.isCompleted() != t2.isCompleted()) return t1.isCompleted() ? 1 : -1;
                        return Integer.compare(t2.id(), t1.id());
                      })
                  .collect(Collectors.toList());

          Map<String, Object> model = new HashMap<>();
          model.put("tasks", finalTasks);
          model.put("view", view);
          model.put("isActive", "active".equals(view));
          model.put("isDrafts", "drafts".equals(view));
          model.put("isTrash", "trash".equals(view));

          return engine.render(new ModelAndView(model, "index.hbs"));
        });

    post(
        "/add",
        (req, res) -> {
          String title = req.queryParams("title");
          boolean isDraft = "on".equals(req.queryParams("is_draft"));
          if (title != null && !title.isEmpty()) {
            int newId = getNextId("tasks", "id");
            String createdAt =
                LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm"));
            db.execute(
                "INSERT INTO tasks VALUES (?, ?, 'pending', ?, ?, ?)",
                Arrays.asList(newId, title, createdAt, isDraft, false));
          }
          res.redirect("/tasks/" + (isDraft ? "drafts" : "active"));
          return null;
        });

    post(
        "/subtask/add",
        (req, res) -> {
          String parentIdStr = req.queryParams("parent_id");
          String subtitle = req.queryParams("subtitle");
          if (parentIdStr != null && subtitle != null) {
            int sid = getNextId("subtasks", "sid");
            db.execute(
                "INSERT INTO subtasks VALUES (?, ?, ?, 'pending')",
                Arrays.asList(sid, Integer.parseInt(parentIdStr), subtitle));
          }
          res.redirect("/tasks/active");
          return null;
        });

    get(
        "/subtask/toggle/:sid",
        (req, res) -> {
          int sid = Integer.parseInt(req.params(":sid"));
          ExecutionResult r =
              db.execute("SELECT * FROM subtasks WHERE sid = ?", Collections.singletonList(sid));
          if (!r.getRows().isEmpty()) {
            String next =
                "pending".equals(r.getRows().get(0).get("substatus")) ? "completed" : "pending";
            db.execute("UPDATE subtasks SET substatus = ? WHERE sid = ?", Arrays.asList(next, sid));
          }
          res.redirect("/tasks/active");
          return null;
        });

    get(
        "/action/:action/:id",
        (req, res) -> {
          String action = req.params(":action");
          int id = Integer.parseInt(req.params(":id"));
          ExecutionResult r =
              db.execute("SELECT * FROM tasks WHERE id = ?", Collections.singletonList(id));
          if (r.getRows().isEmpty()) return "Not Found";
          Map<String, Object> task = r.getRows().get(0);

          if ("toggle".equals(action)) {
            String status = "pending".equals(task.get("status")) ? "completed" : "pending";
            db.execute("UPDATE tasks SET status = ? WHERE id = ?", Arrays.asList(status, id));
          } else if ("delete".equals(action)) {
            db.execute("UPDATE tasks SET is_deleted = ? WHERE id = ?", Arrays.asList(true, id));
          } else if ("restore".equals(action)) {
            db.execute("UPDATE tasks SET is_deleted = ? WHERE id = ?", Arrays.asList(false, id));
          } else if ("publish".equals(action)) {
            db.execute("UPDATE tasks SET is_draft = ? WHERE id = ?", Arrays.asList(false, id));
          } else if ("destroy".equals(action)) {
            db.execute("DELETE FROM tasks WHERE id = ?", Collections.singletonList(id));
            db.execute("DELETE FROM subtasks WHERE parent_id = ?", Collections.singletonList(id));
          }

          String ref = req.headers("Referer");
          if (ref != null && ref.contains("trash")) res.redirect("/tasks/trash");
          else if (ref != null && ref.contains("drafts")) res.redirect("/tasks/drafts");
          else res.redirect("/tasks/active");
          return null;
        });

    System.out.println("Java Server running on http://localhost:4567");
  }

  private static void initDb() {
    List<String> tables = db.getStorage().listTables();
    if (!tables.contains("tasks")) {
      db.execute(
          "CREATE TABLE tasks (id INTEGER PRIMARY KEY, title TEXT, status TEXT, created_at TEXT, is_draft BOOLEAN, is_deleted BOOLEAN)");
    }
    if (!tables.contains("subtasks")) {
      db.execute(
          "CREATE TABLE subtasks (sid INTEGER PRIMARY KEY, parent_id INTEGER, subtitle TEXT, substatus TEXT)");
    }
  }

  private static int getNextId(String table, String idCol) {
    ExecutionResult r = db.execute("SELECT * FROM " + table);
    return r.getRows().stream().mapToInt(row -> (Integer) row.get(idCol)).max().orElse(0) + 1;
  }

  // Helper class to accumulate subtasks
  private static class TaskBuilder {
    private final int id;
    private final String title;
    private final String status;
    private final String createdAt;
    private final boolean isDraft;
    private final boolean isDeleted;
    private final List<SubTask> subtasks = new ArrayList<>();

    TaskBuilder(Map<String, Object> row) {
      this.id = (Integer) row.get("id");
      this.title = (String) row.get("title");
      this.status = (String) row.get("status");
      this.createdAt = (String) row.get("created_at");
      this.isDraft = (Boolean) row.get("is_draft");
      this.isDeleted = (Boolean) row.get("is_deleted");
    }

    void addSubTask(Map<String, Object> row) {
      subtasks.add(
          new SubTask(
              (Integer) row.get("sid"),
              (Integer) row.get("parent_id"),
              (String) row.get("subtitle"),
              (String) row.get("substatus"),
              "completed".equals(row.get("substatus"))));
    }

    Task build() {
      return new Task(
          id, title, status, "completed".equals(status), createdAt, isDraft, isDeleted, subtasks);
    }
  }
}
